src/block-cache.ts

```diff
@@ -0,0 +1,183 @@
+    const cached: unknown = identifier ? blockCache[(identifier as string)] : undefined;
+    // may be undefined
+    return cached;
+  }
+
+  async set(payload: JsonRpcRequest<string[]>, requestedBlockNumber: string, result: Record<string, unknown>): Promise<void> {
+    // check if we can cached this result
+    const canCache: boolean = this.canCacheResult(payload, result);
+    if (!canCache) {
+      return;
+    }
+    // set the value in the cache
+    const blockCache: Record<string, unknown> = this.getBlockCacheForPayload(payload, requestedBlockNumber);
+    const identifier: string|null = cacheUtils.cacheIdentifierForPayload(payload, true);
+    blockCache[(identifier as string)] = result;
+  }
+
+  canCacheRequest(payload: JsonRpcRequest<string[]>,): boolean {
+    // check request method
+    if (!cacheUtils.canCache(payload)) {
+      return false;
+    }
+    // check blockTag
+    const blockTag: string = cacheUtils.blockTagForPayload(payload) as string;
+    if (blockTag === 'pending') {
+      return false;
+    }
+    // can be cached
+    return true;
+  }
+
+  canCacheResult(payload: JsonRpcRequest<string[]>, result: Record<string, unknown>): boolean {
+    // never cache empty values (e.g. undefined)
+    if (emptyValues.includes((result as unknown) as string)) {
+      return false;
+    }
+    // check if transactions have block reference before caching
+    if (['eth_getTransactionByHash', 'eth_getTransactionReceipt'].includes(payload.method)) {
+      if (!result || !result.blockHash || result.blockHash === '0x0000000000000000000000000000000000000000000000000000000000000000') {
+        return false;
+      }
+    }
+    // otherwise true
+    return true;
+  }
+
+  // removes all block caches with block number lower than `oldBlockHex`
+  clearBefore(oldBlockHex: string): void {
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const self = this;
+    const oldBlockNumber: number = Number.parseInt(oldBlockHex, 16);
+    // clear old caches
+    Object.keys(self.cache)
+      .map(Number)
+      .filter((num) => num < oldBlockNumber)
+      .forEach((num) => delete self.cache[num]);
+  }
+}
+
+function createBlockCacheMiddleware(
+  opts: BlockCacheMiddlewareOptions = {},
+): JsonRpcMiddleware<string[], Record<string, unknown>> {
+  // validate options
+  const { blockTracker } = opts;
+
+  if (!blockTracker) {
+    throw new Error('createBlockCacheMiddleware - No BlockTracker specified');
+  }
+
+  // create caching strategies
+  const blockCache: BlockCacheStrategy = new BlockCacheStrategy();
+  const strategies: Record<string, BlockCacheStrategy> = {
+    perma: blockCache,
+    block: blockCache,
+    fork: blockCache,
+  };
+
+  return createAsyncMiddleware(async (req, res, next) => {
+    // allow cach to be skipped if so specified
+    if (((req as unknown) as Record<string, unknown>).skipCache) {
+      return next();
+    }
```

Comment:
I think we can define an interface that supersets the interface expected by `createAsyncMiddleware` for the `req` param that we can cast to on L131, instead of doing current wordy and less specific cast.

Reply:
That could then also be used as the return type for the `JsonRpcMiddleware` return value of `createBlockCacheMiddleware` (i.e. the function we're in).